tensorflow/lite/micro/kernels/cmsis-nn/fully_connected.cc

```diff
@@ -159,9 +161,9 @@ TfLiteStatus EvalQuantizedInt8(TfLiteContext* context, TfLiteNode* node,
 
     cmsis_nn_dims input_dims;
     input_dims.n = batches;
```

Comment:
All the 4 dims are used here to handle cases without reshape before a fully connected op. Something similar to how it is done in  TFL->Arm Ccompute Library. 
In CMSIS-NN, it is treated as having 2 dims  https://github.com/ARM-software/CMSIS_5/blob/develop/CMSIS/NN/Include/arm_nnfunctions.h#L1150 . I'll let Pete or someone from the TFLu team to comment on this as they know best what applies here that would hold true for all cases.

Reply:
Hello @felix-johnny,
sorry for the delay of my reply and thanks for your comment.
You are right and I know that the input tensor is treated as having 2 dims in CMSIS.
However, the point is that when you get the input tensor shape (`const RuntimeShape input_shape = GetTensorShape(input);`), you obtain that the tensor has two dimensions (I checked it through a debugger: `size_` is equal to 2).
So the point is that the mistake is not related to the way the code initializes `cmsis_nn_dims input_dims;` (that is the semantic is fine), but in the way the code uses `input_shape` since the code tries to access to dimensions higher that 2, and thus generating an error (the assertion `TFLITE_DCHECK_LT(i, size_)` becomes false.
I hope I was clear.

See also #41816.

Best regards,
Biagio-

Reply:
Thanks for catching this and making this PR. While various input shapes are valid (including 4-dimensional inputs, as far as I can tell), assuming there are 4 dimensions by accessing input.Dims(2) and input.Dims(3) is certainly incorrect.

I would prefer to make the prepare check ensure that the flat size of the input is equal to batches * accum_depth. The reference kernels do not explicitly check anything about the input dimensions, but they do assume that the flat size is correct. This will mean that instead of using input.Dims(2) and input.Dims(3), input dims would read:
input_dims.n = batches;
input_dims.h = 1; 
input_dims.w = 1;
input_dims.c = accum_depth;

Does this make sense, and do you think this will also fix the issue?

Reply:
Hello @njeffrie,
thanks for your review and sorry for late reply.

Your answer make sense and will fix the issue.
So, is it ok if I update my changes with your suggestion (that is `input_dims.c = input_shape.Dims(1);` -> ` input_dims.c = accum_depth;`) ?

Reply:
Looks good.

Reply:
Hi @njeffrie, I updated the branch on which this pull request is based on with a wrong local branch, so this pull has been closed. I created a new pull request (#42314), which addresses your comments. Sorry for this mistake.

Best regards,
Biagio.